code/assignTime.m

```diff
@@ -0,0 +1,279 @@
+    % seconds)
+    error = expectedElapsed(currentTimestampIndices) - diff_PacketGenTime(currentTimestampIndices);
+    meanError(iChunk) = median(error);
+end
+%%
+% Create corrected timing for each chunk
+counter = 1;
+for iChunk = 1:numChunks
+    if ~ismember(iChunk,chunksToExclude)
+        alignTime = dataTable.PacketGenTime(chunkIndices{iChunk}(1));
+        % alignTime in ms; meanError in units of systemTick
+        correctedAlignTime(counter) = alignTime + meanError(iChunk)*1e-1;
+        % TO DO: add or subtract meanError above??
+        counter = counter + 1;
+    end
+end
+
+%%
+% Indices in chunkIndices correspond to packets in dataTable.
+% CorrectedAlignTime corresponds to first packet for each chunk in
+% chunkIndices. Remove chunks identified above
+chunkIndices(chunksToExclude) = [];
+correctedAlignTime(find(isnan(correctedAlignTime))) = [];
+
+% correctedAlignTime is shifted slightly to keep times exactly aligned to
+% sampling rate; use maxFs for alignment points. In other words, take first
+% correctedAlignTime, and place all other correctedAlignTimes at multiples of
+% (1/Fs)
+deltaTime = 1/maxFs * 1000; % in milliseconds
+
+disp('Shifting chunks to align with sampling rate')
+multiples = floor(((correctedAlignTime - correctedAlignTime(1))/deltaTime) + 0.5);
+correctedAlignTime_shifted = correctedAlignTime(1) + (multiples * deltaTime);
+
+% Full form data table
+outputDataTable = inputDataTable;
+clear inputDataTable
+
+% Remove packets and samples identified above as lacking proper metadata
+samplesToRemove = [];
+% If first packet is included, collect those samples separately
+if ismember(1,packetsToRemove)
+    samplesToRemove = 1:indicesOfTimestamps(1);
+    packetsToRemove = setdiff(packetsToRemove,1);
+end
+% Loop through all other packetsToRemove
+toRemove_start = indicesOfTimestamps(packetsToRemove - 1) + 1;
+toRemove_stop = indicesOfTimestamps(packetsToRemove);
+for iPacket = 1:length(packetsToRemove)
+    samplesToRemove = [samplesToRemove toRemove_start(iPacket):toRemove_stop(iPacket)];
+end
+outputDataTable(samplesToRemove,:) = [];
+
+% Indices referenced in chunkIndices can now be mapped back to timeDomainData
+% using indicesOfTimestamps_cleaned
+indicesOfTimestamps_cleaned = find(outputDataTable.timestamp ~= 0);
+
+% Map the chunk start/stop times back to samples
+for iChunk = 1:length(chunkIndices)
+    currentPackets = chunkIndices{iChunk};
+    chunkPacketStart(iChunk) = indicesOfTimestamps_cleaned(currentPackets(1));
+    if currentPackets(1) == 1 % First packet, thus take first sample
+        chunkSampleStart(iChunk) = 1;
+    else
+        chunkSampleStart(iChunk) = indicesOfTimestamps_cleaned(currentPackets(1) - 1) + 1;
+    end
+    chunkSampleEnd(iChunk) = indicesOfTimestamps_cleaned(currentPackets(end));
+end
+
+disp('Creating derivedTime for each sample')
+% Use correctedAlignTime and sampling rate to assign each included sample a
+% derivedTime
+
+% Initalize DerivedTime
```

Comment:
I just wanted to add a little bit more detail about how the time series for each chunk is created as a comment. If there is any mistake, please add the corrections. Also, I was able to make the loop run a lot faster with my changes.

```suggestion
% Initalize DerivedTime
DerivedTime = nan(size(outputDataTable,1),1);
for iChunk = 1:length(chunkIndices)
    % Display status
    if iChunk > 0 && mod(iChunk, 1000) == 0
        disp(['Currently on chunk ' num2str(iChunk) ' of ' num2str(length(chunkIndices))])
    end
    % Assign derivedTimes to all samples (from before first packet time to end) -- all same
    % sampling rate
    currentFs = outputDataTable.samplerate(chunkPacketStart(iChunk));
    elapsedTime_before = (chunkPacketStart(iChunk) - chunkSampleStart(iChunk)) * (1000/currentFs);
    elapsedTime_after = (chunkSampleEnd(iChunk) - chunkPacketStart(iChunk)) * (1000/currentFs);
    
    DerivedTime(chunkSampleStart(iChunk):chunkSampleEnd(iChunk)) = ...
        correctedAlignTime_shifted(iChunk) - elapsedTime_before : 1000/currentFs : correctedAlignTime_shifted(iChunk) + elapsedTime_after;
end

% All samples which do not have a derivedTime should be removed from final
% data table
disp('Cleaning up output table')
outputDataTable.DerivedTime = DerivedTime;
rowsToRemove = find(DerivedTime == 0);
```

Reply:
Great! I have committed these changes